plugins/buildingplan/buildingplan_cycle.cpp

```diff
@@ -228,84 +228,125 @@ static df::building * popInvalidTasks(color_ostream &out, Bucket &task_queue,
-                        pb.item_filters[rev_filter_idx], pb.specials)
-                    && Job::attachJobItem(job, item,
-                        df::job_item_ref::Hauled, filter_idx))
-            {
+
+            // first task of the bucket: filter/count available items
+            if (first_task) {
+                matching.clear();
+                for(auto item : available)
+                    if (!item->flags.bits.in_job &&
+                        matchesFilters(item,
+                                       jitems[filter_idx],
+                                       pb.heat_safety,
+                                       pb.item_filters[rev_filter_idx],
+                                       pb.specials))
+                        matching.emplace_back(item->pos,item);
+
+                num_matching = matching.size();
+                first_task = false;
+                TRACE(cycle,out).print("first task in bucket: found %zu matching items\n",
+                                        num_matching);
+            }
+            // every task: find and attach closest matching item (if any)
+            if (num_matching == 0)
+                break; // no more items for this bucket, go to next bucket.
+
+            auto jpos = job->pos;
+            std::pair<df::coord, df::item*> *closest = nullptr;
+            for (auto &p : matching) {
+                if (p.second && (closest == nullptr || distance(jpos, p.first) < distance(jpos, closest->first)))
+                    closest = &p;
+            }
+            auto item = closest->second; // some item must be closest.
+
+            if (Job::attachJobItem(job, item, df::job_item_ref::Hauled, filter_idx)) {
                 MaterialInfo material;
                 material.decode(item);
                 ItemTypeInfo item_type;
                 item_type.decode(item);
-                DEBUG(cycle,out).print("attached %s %s to filter %d for %s(%d): %s/%s\n",
+                DEBUG(cycle,out).print("attached %s %s (distance %d) to filter %d for %s(%d): %s/%s\n",
                       material.toString().c_str(),
                       item_type.toString().c_str(),
+                      distance(closest->first, jpos),
                       filter_idx,
                       ENUM_KEY_STR(building_type, bld->getType()).c_str(),
                       id,
                       ENUM_KEY_STR(job_item_vector_id, vector_id).c_str(),
                       bucket_it->first.c_str());
+                // clean up fulfilled task
+                task_queue.pop_front();
                 // keep quantity aligned with the actual number of remaining
                 // items so if buildingplan is turned off, the building will
                 // be completed with the correct number of items.
                 --jitems[filter_idx]->quantity;
-                task_queue.pop_front();
+                // invaldate attached item
+                closest->second = nullptr;
```

Comment:
`closest` goes out of scope in a few lines. moreover, the item it points to is still valid. why invalidate?

Reply:
Evidently, the comment isn't clear. `closest` is a pointer to a `(coord,item*)` cell in `matching`, so it nulls the second component of that, ensuring that this item gets skipped when looking for the closest item for subsequent tasks. Hence the `if (p.second && ...)`

Reply:
Ah ok, I missed that. I conflated it with the `in_job` check for the available vector.